src/utils/futures.rs

```diff
@@ -119,21 +145,68 @@ impl RemoteThread {
 
     /// Create a future in the remote thread and run it.
     ///
-    /// The returned future resolves when the spawned future has completed
-    /// execution.  The output type of the spawned future is wrapped in a
-    /// `Result`, normally the output is returned in an `Ok` but when the remote
-    /// future is dropped because the thread is shut down or restarted for any
-    /// reason it returns `Err(oneshot::Canceled)`.
-    pub fn spawn<F, R, T>(&self, factory: F) -> impl Future<Output = Result<T, RemoteCanceled>>
+    /// The returned future resolves when the spawned future has completed execution.  The
+    /// output type of the spawned future is wrapped in a `Result`, normally the output is
+    /// returned in an `Ok`.
+    ///
+    /// Then the future takes too long `SpawnError::Timeout` is returned, ifthe
+    /// `RemoteThread` it is running on is dropped `SpawnError::Canceled` is returned.
+    pub fn spawn<F, R, T>(
+        &self,
+        task_name: &'static str,
+        timeout: Duration,
+        factory: F,
+    ) -> impl Future<Output = Result<T, SpawnError>>
     where
         F: FnOnce() -> R + Send + 'static,
         R: Future<Output = T> + 'static,
         T: Send + 'static,
     {
+        enum ChannelMsg<T> {
+            Output(T),
+            Timeout,
+        };
         let (tx, rx) = oneshot::channel();
-        let msg = actix::msgs::Execute::new(|| -> Result<(), ()> {
-            let fut = factory().map(|output| tx.send(output).or(Err(())));
-            let fut01 = fut.boxed_local().compat();
+        let creation_time = Instant::now();
+        let msg = actix::msgs::Execute::new(move || -> Result<(), ()> {
+            let fut01 = futures01::future::lazy(move || {
+                metric!(
+                    timer("futures.wait_time") = creation_time.elapsed(),
+                    "task_name" => task_name,
+                );
+                let start_time = Instant::now();
+                factory()
+                    .then(future::ok)
+                    .boxed_local()
+                    .compat()
+                    .timeout(timeout)
+                    .then(move |r: Result<T, tokio::timer::timeout::Error<()>>| {
+                        match r {
+                            // TODO: Should send failures be logged?  Panic?
+                            Ok(o) => {
+                                metric!(
+                                    timer("futures.done") = start_time.elapsed(),
+                                    "task_name" => task_name,
+                                    "status" => "done",
```

Comment:
maybe you can move the duplicate code above, and just use `"status": if r.is_ok() { "done" } else { "timeout" }`

Reply:
good call, done